File: 1696-jump-game-vi/1696-jump-game-vi.py
class Solution:
    def maxResult(self, nums: List[int], k: int) -> int:
        n = len(nums)
        
        ans = nums[-1]
        max_heap = [(-nums[-1], n - 1)]
        for i in reversed(range(n - 1)):
            while max_heap and max_heap[0][1] > i + k:
                heappop(max_heap)
            ans = nums[i] - max_heap[0][0]
            heappush(max_heap, (-ans, i))
        return ans

File: 1696-jump-game-vi/1696-jump-game-vi.java
class Solution {
    public int maxResult(int[] nums, int k) {
        LinkedList<Pair<Integer,Integer>> q = new LinkedList<>();
        int ans = nums[nums.length - 1];
        
        q.add(new Pair(ans, nums.length - 1));
        for(int i = nums.length - 2; i >= 0; i--){
            while(q.size() > 0 && q.getFirst().getValue() > i + k)
                q.removeFirst();
            
            ans = nums[i] + q.getFirst().getKey();
            
            while(q.size() > 0 && q.getLast().getKey() < ans)
                q.removeLast();
            q.add(new Pair(ans, i));
        }
        
        return ans;
    }
}